homeassistant/components/control4/__init__.py

```diff
@@ -135,30 +98,161 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Return a list of all Control4 items with the specified category."""
-    director_all_items = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR_ALL_ITEMS]
-    return_list = []
-    for item in director_all_items:
-        if "categories" in item and category in item["categories"]:
-            return_list.append(item)
-    return return_list
+    _LOGGER.debug("Getting items of category: %s", category)
+    director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+    return_list = await director.getAllItemsByCategory(category)
+    return json.loads(return_list)
+
 
+async def refresh_tokens(hass: HomeAssistant, entry: ConfigEntry):
+    """Store updated authentication and director tokens in hass.data, and schedule next token refresh."""
+    config = entry.data
+    verify_ssl_session = aiohttp_client.async_get_clientsession(hass)
 
-class Control4Entity(CoordinatorEntity):
+    account = C4Account(
+        config[CONF_USERNAME], config[CONF_PASSWORD], verify_ssl_session
+    )
+    try:
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise ConfigEntryNotReady from exception
+    except BadCredentials as exception:
+        _LOGGER.error(
+            "Error authenticating with Control4 account API, incorrect username or password: %s",
+            exception,
+        )
+        raise ConfigEntryAuthFailed(exception) from exception
+
+    controller_unique_id = config[CONF_CONTROLLER_UNIQUE_ID]
+    director_token_dict = await account.getDirectorBearerToken(controller_unique_id)
+    no_verify_ssl_session = aiohttp_client.async_get_clientsession(
+        hass, verify_ssl=False
+    )
+
+    director = C4Director(
+        config[CONF_HOST], director_token_dict[CONF_TOKEN], no_verify_ssl_session
+    )
+
+    _LOGGER.debug("Saving new account and director tokens in hass data")
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    entry_data[CONF_ACCOUNT] = account
+    entry_data[CONF_DIRECTOR] = director
+
+    if not (
+        CONF_WEBSOCKET in entry_data
+        and isinstance(entry_data[CONF_WEBSOCKET], C4Websocket)
+    ):
+        _LOGGER.debug("First time setup, creating new C4Websocket object")
+        connection_tracker = C4WebsocketConnectionTracker(hass, entry)
+        websocket = C4Websocket(
+            config[CONF_HOST],
+            no_verify_ssl_session,
+            connection_tracker.connect_callback,
+            connection_tracker.disconnect_callback,
+        )
+        entry_data[CONF_WEBSOCKET] = websocket
+
+        # Silence C4Websocket related loggers, that would otherwise spam INFO logs with debugging messages
+        logging.getLogger("socketio.client").setLevel(logging.WARNING)
+        logging.getLogger("engineio.client").setLevel(logging.WARNING)
+        logging.getLogger("charset_normalizer").setLevel(logging.ERROR)
```

Comment:
We shouldn't change the global loggers here as it will override what the user has asked for

Reply:
I agree that this solution is not optimal. 

The reason I did this is because the third party library I'm using to handle Websockets (python-socketio) writes all its _debug_ logging to the INFO log level. If we don't change these global loggers, Home Assistant INFO logs would be spammed with multiple messages every second.

I tried to search for a way to redirect all that verbose logging to the DEBUG log level, but I couldn't find a way to do so.

Reply:
You can use a filter instead like the `mjpeg` integration does.